feat: Add profile-specific configuration for disallowed methods and types

Lallapallooza opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/15779)*

Add profile-scoped disallow lists for methods and types, wiring the new configuration tables through a shared resolver that can be toggled with #[clippy::disallowed_profile] attributes.

- extend clippy_config::Conf to parse [disallowed-methods-profiles.*] and [disallowed-types-profiles.*] tables through reusable helpers
- introduce clippy_utils::disallowed_profiles to resolve #[clippy::disallowed_profile(s)] attributes with caching, diagnostics for bad inputs, and shared symbol registrations
- teach DisallowedMethods and DisallowedTypes to honor active profiles, reuse the combined profile set, and annotate diagnostics with the triggering profile name
- add UI TOML coverage for per-profile behavior and refresh the unknown-key fixture for the new configuration entries
- register the new attributes and symbols so tool attributes remain recognized and pedantic lints stay clean

changelog: [disallowed_methods]: allow selecting per-scope disallow lists via disallowed-methods-profiles and the clippy::disallowed_profile attribute
changelog: [disallowed_types]: allow selecting per-scope disallow lists via disallowed-types-profiles and the clippy::disallowed_profile attribute

Lallapallooza committed:
Add profile-specific configuration for disallowed methods and types

blyxyas reviewed (commented):
I'll recount what I understand here, so a user uses a special configuration for settings profiles in their `clippy.toml` so any function can opt-in to those profiles, enabling warnings to those designated disallows in that function.

For example, one might want to disallow functions that may execute arbitrary code from a string input, into user-input-handling functions, and this is a way to achieve via Clippy warning of those uses.

Am I correct, have I forgotten/misunderstood something? If that's the case, we might be interested in generalizing this a bit so that both (1.) we can use profiles for other lints without the need of adding a special attribute for each lint, and (2.) people can use the general idea of `profiles` in Clippy to segment their source code for several lints at the same time.

Also, @ojeda could this be interesting for the Rust4Linux project? Should I give it priority?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15779/3c93ba0c1dd456e0da4ada1e54fd5df3e6b930d1..a8c3a91a1b2eda78f03ddbb69f958c152062134c)*

blyxyas commented:
Hi Lallapallooza, I've talked with the team and I'm going to proceed with the review. Also, sorry for the delayed response, I've just returned from vacation.

I was thinking that this may be useful for this [message thread](https://lore.kernel.org/rust-for-linux/5c0f7b5f-ae57-4a9e-9aa2-9628a7bc6a2c@nvidia.com/#t)

> After learning about disallowed_names ("foo", "bar" and others that are
in fact classical documentation favorites), I also think it would be
very nice to allow those in documentation.

 I thought that we already had a configuration option for this, but it seems that we don't have one. So I think that this feature can also support this use case.

Lallapallooza committed:
fix review

blyxyas commented on clippy_lints/src/disallowed_methods.rs:
```diff
@@ -55,19 +60,39 @@ declare_clippy_lint! {
+    /// [disallowed-methods-profiles.forward_pass]
+    /// paths = [
+    ///     { path = "crate::devices::Buffer::copy_to_host", reason = "Forward code must not touch host buffers" }
+    /// ]
```
(See other comments about API and more general examples)

```suggestion
    /// [profiles.single_threaded]
    /// disallowed_methods = [
    ///     { path = "std::thread::spawn", reason = "Unsafe to spawn more threads here" }
    /// ]
```

blyxyas reviewed (changes requested):
We'll need some interface changes to be more ergonomic, but apart from that, this looks great!

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15779/63c7197be0ed379d840ec7a444297ba2a728dd6f..0ff61e72ef9e2f44a92523c393b7f0dc52bce662)*

Lallapallooza committed:
config: add [profiles.*] disallowed lists

blyxyas commented on clippy_lints/src/disallowed_types.rs:
```diff
@@ -51,43 +54,144 @@ declare_clippy_lint! {
+        let default = TypeLookup::from_config(tcx, &conf.disallowed_types);
+
+        let mut profiles = FxHashMap::default();
+        let mut names: Vec<_> = conf.profiles.keys().collect();
```
Same as with `disallowed_methods`

blyxyas commented on tests/ui-toml/disallowed_profiles_methods/main.rs:
We also need some testing with `expect`s and unknown-profile error messages. Spans tend to mess those up and if we're not careful how they're handled, it's really messy later on.

blyxyas reviewed (changes requested):
Some more thoughts ฅ(´・ω・｀)ฅ

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15779/63c7197be0ed379d840ec7a444297ba2a728dd6f..af37815091ab38bb2f69b26627bc6f15b7cf636e)*

Lallapallooza committed:
fix: address review feedback

Lallapallooza commented on clippy_config/src/conf.rs:
```diff
@@ -221,12 +222,80 @@ macro_rules! deserialize {
+    };
+}
+
+macro_rules! deserialize_profiles {
```
removed

Lallapallooza commented on clippy_lints/src/disallowed_methods.rs:
```diff
@@ -80,7 +103,66 @@ impl DisallowedMethods {
+                        entry.name
+                    ),
+                )
+                .emit();
```
replaced to `span_lint`

blyxyas commented on tests/ui-toml/disallowed_profiles_methods/main.rs:
```diff
@@ -12,6 +12,12 @@ fn default_violation() {
     std::mem::drop(value); //~ ERROR: use of a disallowed method `std::mem::drop`
 }
 
+#[expect(clippy::disallowed_methods)]
```
We also need expects for the `unknown_profile` lint, and trying it before and after the `profile` attribute.

Lallapallooza replied:
Added tests in both `disallowed_profiles_methods` and `disallowed_profiles_types` covering `#[expect(...)]` placed before and after `#[clippy::disallowed_profile("unknown_...")]`.

blyxyas commented on clippy_lints/src/disallowed_types.rs:
```diff
@@ -51,43 +54,139 @@ declare_clippy_lint! {
+        }
+    }
+}
+
+pub struct DisallowedTypes {
+    default: TypeLookup,
+    profiles: FxHashMap<Symbol, TypeLookup>,
+    known_profiles: FxHashSet<Symbol>,
```
Why is `profiles` different from `known_profiles`? Is there a difference between these two? Seems that `known_profiles` is just a set from the keys of `profiles`.

Lallapallooza replied:
They're intentionally different: `known_profiles` holds every name declared in `[profiles.*]`, while `profiles` only holds those that contribute entries to *this* lint.

The distinction is what suppresses an "unknown profile" warning when a profile only defines entries for the other lint.

Added field-level comments to make this explicit. Thanks

Lallapallooza committed:
fix: address final review feedback

- document `ProfileEntry` and `ProfileSelection`
- explain why `profiles` and `known_profiles` are separate
- emit unknown-profile warning inline (drop intermediate SmallVec)
- add tests covering `#[expect]` before/after `#[clippy::disallowed_profile]`
  with an unknown profile name, for both methods and types lints
- sort_by -> sort_by_key to satisfy dogfood

Lallapallooza commented on clippy_utils/src/disallowed_profiles.rs:
```diff
@@ -0,0 +1,169 @@
+use rustc_span::{Span, Symbol};
+
+#[derive(Copy, Clone)]
+pub struct ProfileEntry {
```
Added a doc comment explaining that each entry represents one profile-name argument from a `#[clippy::disallowed_profile(s)]` attribute. Thanks!

Lallapallooza commented on clippy_utils/src/disallowed_profiles.rs:
```diff
@@ -0,0 +1,169 @@
+}
+
+#[derive(Clone)]
+pub struct ProfileSelection {
```
Added a doc comment, it's the set of profiles resolved for a given `HirId` by walking the attribute-bearing HIR ancestors.

Lallapallooza commented on clippy_lints/src/disallowed_types.rs:
```diff
@@ -51,43 +54,139 @@ declare_clippy_lint! {
+    }
+
+    fn check_res_emit(&mut self, cx: &LateContext<'_>, hir_id: rustc_hir::HirId, res: &Res, span: Span) {
+        let mut active_profiles = SmallVec::<[Symbol; 2]>::new();
```
Thank you! 🙏

Lallapallooza commented on clippy_lints/src/disallowed_types.rs:
```diff
@@ -51,43 +54,139 @@ declare_clippy_lint! {
+
+        for entry in unknown_profiles {
+            self.warn_unknown_profile(cx, &entry);
+        }
```
Done, thanks! Dropped the intermediate `unknown_profiles` `SmallVec` and warn inline.

I had to copy entries out of the cache first because `warn_unknown_profile` takes `&mut self` and the `active_profiles` borrow holds `&mut self.profile_cache`.

Lallapallooza commented on clippy_lints/src/disallowed_types.rs:
```diff
@@ -51,43 +54,139 @@ declare_clippy_lint! {
+                    unknown_profiles.push(*entry);
+                }
+            }
+        }
```
Applied the inline-warn refactor from the sibling thread. Kept the `known_profiles` guard so that profiles used for the other lint don't get flagged as unknown here.

Lallapallooza commented:
Hi @blyxyas, sorry for leaving this PR sitting for so long.

I've rebased onto current master and addressed all the outstanding review comments (docs on `ProfileEntry`/`ProfileSelection`, the `profiles` vs `known_profiles` distinction, inlining the unknown-profile warning, and the `#[expect]` before/after tests).

Replies are in each thread. Whenever you have a moment, could you take another look? Thanks again for the reviews. 🙏

Lallapallooza commented:
@blyxyas Hi, could you check my patch again please.

blyxyas reviewed (approved):
LGTM, thanks! :heart:
Thanks for taking this amazing contribution until the end, I think that it will heavily improve workflows, and there are many people interested in this kind of configuration option. I might make a blog post about it.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15779/f058fe788d09e3f517848ec06cc3e23025ec61d5..647bc062714e1b0898b3e82088a5ab2ad5555919)*

Lallapallooza commented:
@blyxyas Thanks! I rebased MR and now it's mergeable.

Jarcho commented:
Why did this not get an FCP? I remember there being a zulip thread quite a while ago that had some objections that got left hanging.

blyxyas commented:
I'll revert this pull request, I forgot about those objections (I'll look them up right now). I'm sorry for all nightly users that got an un-FCP'd feature in their nightly.

Lallapallooza commented:
@blyxyas @Jarcho Could you link the Zulip thread with the objections? I'd like to address them so this can go through FCP properly and get re-merged.

blyxyas commented:
The thread is here: https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/Profiles.20for.20per-lint.20configuration/with/563940955

Jarcho commented:
The objection was more a vague vibes based objection, not any specific issues. Not an insurmountable thing in any way, but it should've been addressed before this was merged.

Jarcho commented:
This has now been reverted. Again, this isn't rejection of the feature in some form (or even the current form). The issue here is that this is a notable config change/lint extension which had some concerns raised which weren't addressed before it was merged.

Lallapallooza commented:
> This has now been reverted. Again, this isn't rejection of the feature in some form (or even the current form). The issue here is that this is a notable config change/lint extension which had some concerns raised which weren't addressed before it was merged.

Thanks. Please let me know after the FCP if any changes are required from me. I’m happy to adjust the patch based on the requirements.

Jarcho commented:
Can you open a new PR please. It makes tracking things easier if the FCP is linked to an open PR.

Lallapallooza commented:
> Can you open a new PR please. It makes tracking things easier if the FCP is linked to an open PR.

https://github.com/rust-lang/rust-clippy/pull/17301
